ref/src/interpolate.f90

```diff
@@ -0,0 +1,180 @@
+    !
+    ! f's size has been calculated in subroutine interpolate_allocate
+    ! f has been allocated in the call tree above this subroutine
+    !
+    ! Subroutine Parameters:
+    ! oa          - original array, 2D
+    ! odims       - low/high subscripts for oa
+    ! f           - interpolated array, 2D
+    ! fdims       - low/high subscripts for f 
+    ! interpFactor- interpolation factor
+
+    integer, intent(in)    :: odims(4) ! 
+    real,    intent(in)    :: oa(odims(1):odims(2), odims(3):odims(4))
+    integer, intent(in)    :: fdims(4)
+    real,    intent(inout) :: f(fdims(1):fdims(2), fdims(3):fdims(4))
+    integer, intent(in)    :: interpFactor
+
+    ! Here are some illustrations of interpFactor.
+    ! We are assuming that the grid points are equally spaced.
+    !
+    ! An interpolation factor of 1 means 1 new interpolated element.
+    ! So for example, a 3x3 matrix with interpFactor=1 becomes a 5x5 matrix
+    !
+    ! x's are data from oa, the o's are points to be interpolated 
+    !
+    ! [ x x x ]     ==> [x o x o x]
+    ! [ x x x ]         [o o o o o]
+    ! [ x x x ]         [x o x o x]
+    !                   [o o o o o]
+    !                   [x o x o x]
+    !
+    ! And an interpFactor=2, means a 3x3 matrix becomes a 7x7 matrix.
+    ! [ x x x ]     ==> [x o o x o o x]
+    ! [ x x x ]         [o o o o o o o]
+    ! [ x x x ]         [o o o o o o o]
+    !                   [x o o x o o x]
+    !                   [o o o o o o o]
+    !                   [o o o o o o o]
+    !                   [x o o x o o x]
+
+    ! Locals
+    integer :: x1,x2,y1,y2         ! indexes of the interpolated point
+    integer :: ix,iy               ! indexes of the interpolated point
+    real    :: a0,a1,a2,a3         ! coefficients
+    integer :: icount              ! 1D counter
+    real    :: fQ11,fQ12,fQ21,fQ22 ! function values at the corners
+
+    integer :: ol1,ou1,ol2,ou2     ! lower and upper bounds of 2D array oa
+    integer :: fl1,fu1,fl2,fu2     ! lower and upper bounds of 2d array f
+    integer :: i,j                 ! do loop indexes to loop over the squares
+
+    ! Unlock the subscript bounds of the originalArray.
+    ol1 = odims(1)
+    ou1 = odims(2)
+    ol2 = odims(3)
+    ou2 = odims(4)
+
+    ! Unlock the subscript bounds of the interpolatedArray.
+    fl1 = fdims(1)
+    fu1 = fdims(2)
+    fl2 = fdims(3)
+    fu2 = fdims(4)
+
+    ! Intersperse the oa points into f.
+    icount = 0
+    do j = fl2, fu2, interpFactor + 1
+      do i = fl1, fu1, interpFactor + 1
+        f(i, j) = oa(ol1 + modulo(icount, ou1-ol1+1), &
+                                 (ol2 + icount/(ou1-ol1+1)))
+        icount = icount + 1
+      enddo
+    enddo
+
+    ! Loop over all the squares.
+    do j = fl2, fu2-1, interpFactor + 1
+      do i = fl1, fu1-1, interpFactor + 1
+        ! Loop over all the interpolated points
+        do iy=j,j+interpFactor +1
+           do ix=i,i+interpFactor +1
```

Comment:
Perhaps you could only loop over the interpolated points by:

```
do iy = j + 1, j + interpFactor
  do ix = i + 1, i + interpFactor
```

And then you can eliminate the check for the corners below.

Reply:
Oops.  Nevermind.  I just figure it out why you did it that way.  Sorry.